models/demos/utils/llm_demo_utils.py

```diff
@@ -0,0 +1,114 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+from loguru import logger
+from models.perf.benchmarking_utils import BenchmarkProfiler, BenchmarkData
+
+
+def create_benchmark_data(profiler: BenchmarkProfiler, measurements: dict, N_warmup_iter: dict, targets: dict):
+    """
+    Create a benchmark data object and populate the object with the given measurements.
+
+    Pre-requisites:
+    - The measurements dictionary should contain the following keys: "compile_prefill", "compile_decode", "prefill_t/s", "decode_t/s", "decode_t/s/u"
+    - The profiler object should contain the start and end times for the steps "compile_prefill", "compile_decode", "inference_prefill", "inference_decode"
+
+    Optional (should be provided if measuring perf, not required for token verification):
+    - The targets dictionary should contain the following keys: "prefill_t/s", "decode_t/s", "decode_t/s/u"
+    - The N_warmup_iter dictionary should contain the following keys: "inference_prefill", "inference_decode"
+    """
+
+    assert all(
+        key in measurements
+        for key in [
+            "compile_prefill",
+            "compile_decode",
+            "prefill_t/s",
+            "decode_t/s",
+            "decode_t/s/u",
+        ]
+    )
+
+    benchmark_data = BenchmarkData()
+
+    # Add measurement data
+    benchmark_data.add_measurement(profiler, 0, "compile_prefill", "time(s)", measurements["compile_prefill"])
+    benchmark_data.add_measurement(profiler, 0, "compile_decode", "time(s)", measurements["compile_decode"])
+    benchmark_data.add_measurement(
+        profiler,
+        0,
+        "inference_prefill",
+        "tokens/s",
+        measurements["prefill_t/s"],
+        step_warm_up_num_iterations=N_warmup_iter["inference_prefill"]
+        if "inference_prefill" in N_warmup_iter
+        else None,
+        target=targets["prefill_t/s"] if "prefill_t/s" in targets else None,
+    )
+    benchmark_data.add_measurement(
+        profiler,
+        0,
+        "inference_decode",
+        "tokens/s",
+        measurements["decode_t/s"],
+        step_warm_up_num_iterations=N_warmup_iter["inference_decode"] if "inference_decode" in N_warmup_iter else None,
+        target=targets["decode_t/s"] if "decode_t/s" in targets else None,
+    )
+    benchmark_data.add_measurement(
+        profiler,
+        0,
+        "inference_decode",
+        "tokens/s/user",
+        measurements["decode_t/s/u"],
+        step_warm_up_num_iterations=N_warmup_iter["inference_decode"] if "inference_decode" in N_warmup_iter else None,
+        target=targets["decode_t/s/u"] if "decode_t/s/u" in targets else None,
+    )
+    return benchmark_data
+
+
+def verify_tokens(generated_text: dict, expected_greedy_output_path: str):
```

Comment:
Are we planning to use this in more LLMs? Did you confirm with Models team?

Reply:
Yes, the token and perf verification functions were added here so that other model teams could easily integrate them in their demos.